File: fszmq.props/fszmq.props.vb/Properties.vb
Imports FsCheck.PropertyExtensions
Imports CheckProperty = FsCheck.NUnit.PropertyAttribute

Namespace fszmq.props

  <TestFixture()>
  Public Class Properties

    ''' <summary>
    ''' Tests is 2 Message instances have the same size and data
    ''' </summary>
    Function HasEqualContent (ByVal msg1 As Message, 
                              ByVal msg2 As Message) As Boolean
      Return  msg1.Size() = msg2.Size() AndAlso 
              msg1.Data().SequenceEqual(msg2.Data())
    End Function

    ''' <summary>
    ''' Successive clones should not change the content of a Message
    ''' </summary>
    <CheckProperty(Arbitrary := { GetType(Generators) })>
    Function CloneIsIdempotent (ByVal msg As Message) As Boolean
      Using once  = msg.Clone(), 
            twice = msg.Clone().Clone()

        Return HasEqualContent(once, twice)
      End Using
    End Function
    
    ''' <summary>
    ''' Compare content before and after cloning
    ''' </summary>
    <CheckProperty(Arbitrary := { GetType(Generators) })>
    Function CopyAltersTargetNotSource_Naive (ByVal msg1 As Message,
                                              ByVal msg2 As Message) As Boolean
      Dim source = msg1.Data()
      Dim before = msg2.Data()
    
      msg1.Copy(msg2)
      Dim after = msg2.Data()
    
      Return after.SequenceEqual(source) AndAlso Not after.SequenceEqual(before)
    End Function

    ''' <summary>
    ''' Compare content before and after cloning
    ''' </summary>
    <CheckProperty(Arbitrary := { GetType(Generators) })>
    Function CopyAltersTargetNotSource_Labels (ByVal msg1 As Message,
                                               ByVal msg2 As Message) As [Property]
      Dim source = msg1.Data()
      Dim before = msg2.Data()
    
      msg1.Copy(msg2)
      Dim after = msg2.Data()
      
      Return after.SequenceEqual(source)                  _
                  .Label("Source")                        _
                  .And  (Not after.SequenceEqual(before)) _
                  .Label("Target")
    End Function
  
    ''' <summary>
    ''' Compare dissimilar content before and after cloning
    ''' </summary>
    <CheckProperty(Arbitrary := { GetType(Generators) })>
    Public Function CopyAltersTargetNotSource_Filtered () As [Property]
      Return Prop.ForAll(New Action(Of Message,Message)(Sub (ByVal msg1, ByVal msg2)
        Dim source = msg1.Data()
        Dim before = msg2.Data()
        
        Dim act As Func(Of Boolean) = Function ()
          msg1.Copy(msg2)
          Return msg2.Data().SequenceEqual(source)
        End Function
                                                            
        act.When  (Not before.SequenceEqual(source))      _
           .Label ("Source")                              _
           .And   (Not msg2.Data().SequenceEqual(before)) _
           .Label ("Target")
      End Sub))
      
    End Function

  End Class

End Namespace

File: fszmq.props/fszmq.props.cs/Properties.cs
namespace fszmq.props
{
  using NUnit.Framework;
  using FsCheck;
  using System;
  using System.Linq;

  using CheckProperty = FsCheck.NUnit.PropertyAttribute;

  using fszmq;
  
  [TestFixture]
  public class Properties
  {
    /// <summary>
    /// Tests is 2 Message instances have the same size and data
    /// </summary>
    Boolean hasEqualContent (Message msg1, Message msg2)
    {
      return (msg1.Size() == msg2.Size()) 
          && msg1.Data().SequenceEqual(msg2.Data());
    }

    /// <summary>
    /// Successive clones should not change the content of a Message
    /// </summary>
    [CheckProperty(Arbitrary = new []{ typeof(Generators) })]
    public Boolean CloneIsIdempotent (Message msg)
    {
      using (var once   = msg.Clone())
      using (var twice  = msg.Clone().Clone())
      {
        return hasEqualContent(once, twice);
      }
    }

    /// <summary>
    /// Compare content before and after cloning
    /// </summary>
    [CheckProperty(Arbitrary = new []{ typeof(Generators) })]
    public Boolean CopyAltersTargetNotSource_Naive (Message msg1, Message msg2) 
    {
      var source = msg1.Data();
      var before = msg2.Data();
    
      msg1.Copy(msg2);
      var after = msg2.Data();
    
      return after.SequenceEqual(source) && !after.SequenceEqual(before);
    }

    /// <summary>
    /// Compare content before and after cloning
    /// </summary>
    [CheckProperty(Arbitrary = new []{ typeof(Generators) })]
    public Property CopyAltersTargetNotSource_Labels (Message msg1, Message msg2)
    {
      var source = msg1.Data();
      var before = msg2.Data();

      msg1.Copy(msg2);
      var after = msg2.Data();

      return after.SequenceEqual(source)
                  .Label("Source")
                  .And  (!after.SequenceEqual(before))
                  .Label("Target");
    }

    /// <summary>
    /// Compare dissimilar content before and after cloning
    /// </summary>
    [CheckProperty(Arbitrary = new []{ typeof(Generators) })]
    public Property CopyAltersTargetNotSource_Filtered ()
    {
      return Prop.ForAll<Message,Message>((msg1, msg2) => {
        var source = msg1.Data();
        var before = msg2.Data();
        
        Func<Boolean> act = () => 
        { 
          msg1.Copy(msg2);
          return msg2.Data().SequenceEqual(source);
        };

        act.When  (!before.SequenceEqual(source))
           .Label ("Source")
           .And   (!msg2.Data().SequenceEqual(before))
           .Label ("Target");
      });
    }
  }
}
